ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1443,6 +1470,38 @@ predicate readStepCommon(Node node1, ContentSet c, Node node2) {
   node2 = node1.(SynthHashSplatParameterNode).getAKeywordParameter(c)
   or
   node2 = node1.(SynthSplatParameterNode).getAParameter(c)
+  or
+  // TODO: convert into the above form
+  synthSplatArgumentElementReadStep(node1, c, node2)
+  or
+  // read from SynthSplatArgParameterNode[n] to nth positional parameter
+  exists(SynthSplatArgParameterNode argParamNode, NormalParameterNode posNode, int n |
+    argParamNode = node1 and
+    posNode = node2 and
+    posNode
+        .isParameterOf(argParamNode.getEnclosingCallable(),
+          any(ParameterPosition p | p.isPositional(n))) and
+    c = getPositionalContent(n)
+  )
+}
+
+// read from splat arg to synth splat arg element
+predicate synthSplatArgumentElementReadStep(
+  Node node1, ContentSet c, SynthSplatArgumentElementNode node2
+) {
+  exists(int n, int splatPos, CfgNodes::ExprNodes::CallCfgNode call |
+    // Don't propagate taint on the `self` element of the splat
+    // since that won't (probably) won't reach the parameters of the callable.
+    // This saves a node per call.
+    n >= 0 and
+    node1.asExpr().(Argument).isArgumentOf(call, any(ArgumentPosition p | p.isSplat(splatPos))) and
+    node2.getCall() = call and
+    node2.getPosition() = n + splatPos and
+    (
+      c = getPositionalContent(n) or
+      c.isSingleton(TUnknownElementContent())
```

Comment:
`n` is not bound in this case; in this case it should store into a version of `SynthSplatArgumentElementNode` that represents unknown content (you could abuse `TSynthSplatArgumentElementNode(-1)` to represent unknown).